File: Chapter-6-Constructor-and-Destructor-Functions/002-a-sample-program-to-introduce-multiple-constructors-in-a-class.cpp
// a sample program to introduce multiple constructors in a class
#include <bits/stdc++.h>

using namespace std;

class complexNumber /// x + yj form
{
    float x; /// real part
    float y; /// imaginary part
public:
    complexNumber() {}
    complexNumber(float a) { x = y = a; }
    complexNumber(float real, float imaginary)
    {
        x = real;
        y = imaginary;
    }

    // Friend functions
    friend complexNumber sum(complexNumber, complexNumber);
    friend void showComplexNumber(complexNumber c);
};

complexNumber sum(complexNumber one, complexNumber two)
{
    complexNumber add;
    add.x = one.x + two.x;
    add.y = one.y + two.y;
    return add;
}

void showComplexNumber(complexNumber c)
{
    cout << c.x << " + " << c.y << "j" << endl;
}

int main()
{
    complexNumber A(2.7, 3.5);
    complexNumber B(1.6);
    complexNumber C;
    C = sum(A, B);

    cout << "A = ";
    showComplexNumber(A);
    cout << "B = ";
    showComplexNumber(B);
    cout << "C = ";
    showComplexNumber(C);

    return 0;
}

// output
// A = 3.1 + 5.65j
// B = 2.75 + 1.2j
// C = 5.85 + 6.85j

File: Chapter-6-Constructor-and-Destructor-Functions/002-a-sample-program-to-introduce-multiple-constructors-in-a-class.py
# The Python code mirrors this structure with the __init__ method, 
# which serves as the constructor in Python. 
# The default values are provided for parameters to achieve the behavior of multiple constructors.
class ComplexNumber:
    def __init__(self, real=0.0, imaginary=0.0):
        self.x = real # real part
        self.y = imaginary # imaginary part

def add_complex_numbers(one, two):
    add = ComplexNumber()
    add.x = one.x + two.x
    add.y = one.y + two.y
    return add

def show_complex_number(c):
    print(f"{c.x} + {c.y:.2f}i")

# Creating complex numbers
one = ComplexNumber(3.1, 5.65)
two = ComplexNumber(1.6)

# Adding two complex numbers
C = add_complex_numbers(one, two)

# Displaying the result
print("A = ", end="")
show_complex_number(one) # 3.1 + 5.65i
print("B = ", end="")
show_complex_number(two) # 1.6 + 0.00i
print("C = ", end="")
show_complex_number(C) # 5.7 + 5.65i